homeassistant/components/lorawan/sensor.py

```diff
@@ -0,0 +1,128 @@
+"""Support for MQTT sensors."""
+from __future__ import annotations
+
+import json
+import logging
+
+from pyliblorawan.models import Uplink
+from pyliblorawan.network_servers.ttn import TTN
+
+from homeassistant.components.mqtt.models import ReceiveMessage
+from homeassistant.components.sensor import SensorEntity, SensorStateClass
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import UndefinedType
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DOMAIN, TTN_TOPIC
+from .devices.browan import HassTBMS100
+from .models import SensorTypes
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT sensor through YAML and through MQTT discovery."""
+
+    await _async_setup_entity(hass, async_add_entities, config_entry)
+
+
+async def _async_setup_entity(
+    hass: HomeAssistant,
+    async_add_entities: AddEntitiesCallback,
+    config_entry: ConfigEntry,
+) -> None:
+    """Set up LoRaWAN sensor."""
+    entities = []
+    coordinator = LorawanSensorCoordinator(hass, config_entry)
+    for sensor in HassTBMS100.supported_sensors():
+        entities.append(LorawanSensorEntity(hass, config_entry, coordinator, sensor))
+    async_add_entities(entities)
+    await coordinator.subscribe()
+
+
+class LorawanSensorCoordinator(DataUpdateCoordinator):
+    """Allows to update all entities with one payload reception."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+    ) -> None:
+        """Initialize the coordinator."""
+        self._config = config_entry
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"LorawanSensorCoordinator.{config_entry.title}",
+        )
+
+    async def _message_received(self, msg: ReceiveMessage) -> None:
+        """Handle uplink, parse and normalize it."""
+        uplink = json.loads(msg.payload)
+        uplink = TTN.normalize_uplink(uplink)
+        uplink = await HassTBMS100.parse_uplink(uplink)
+
+        self.async_set_updated_data(uplink)
+
+    async def subscribe(self) -> None:
+        """Subscribe to MQTT messages and handle them."""
+        topic = TTN_TOPIC.replace("<DEVICE_ID>", self._config.title)
+        await self.hass.components.mqtt.async_subscribe(topic, self._message_received)
```

Comment:
I feel like this integration is taking a wrong approach.

- The MQTT integration should be used to make lorawan data accessible in Home Assistant. It shouldn't be wrapped in a config flow for a lorawan integration.
- Each manufacturer should have its own integration. So there should be one for all Browan devices.

Reply:
@balloob thanks, but would you mind to elaborate? If I understand your comment:

- Each LoRaWAN device manufacturer would have its own integration (granted, this would actually be cleaner)
- Each integration/brand would use the MQTT integration. How?
  1. Like I did here by subscribing to topics?
  2. By inheriting from `MqttSensor` instead of `SensorEntity` ?
  3. Where do we decode the binary payload? In this proposition a coordinator does it and dispatch it to each entity attached to the device, as one message from a device usually contain data for several entities
  4. Or instead of ii. and iii. keep the current approach (inherit from SensorEntiry+coordinator) but duplicate it for each manufacturer?
- In such case we need to configure the network servers for every manufacturer, or is there a way to share the configuration between integrations?

network servers: In short it is a server managing the security, addressing, radio parameters of a LoRaWAN network, and delivering the data the application server (Hass in our case). We need to be able to connect to several to benefits from local network server (local devices) and/or cloud (cloud hosted network server, remote devices or trackers).

Reply:
Hey @balloob, would you mind sharing an update on this or show me an integration where I could understand your comment?

Thanks

Reply:
Hello @balloob, I am sure that you have many things to do but I still need some guidance, please help me to understand:

- How to use MQTT integration here as binary payload from the device must be decoded?
- I am OK to create one integration per device manufacturer instead of one LoRaWAN integration but some manufacturers are not doing only LoRaWAN devices, it might create confusion

Thanks